### PR TITLE
Feature/plonk

### PR DESCRIPTION
**Pull request by Hrom131**
- [ ] Since this PR suggests a **bug fix**, the relevant tests have been added.
- [x] Since this PR introduces a **new feature**, the update has been discussed in an Issue or with the team.
- [ ] This PR is just a **minor change**, like a typo fix.

---

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -59,8 +59,8 @@
   },
   "dependencies": {
     "@distributedlab/circom-parser": "0.2.0",
-    "@solarity/zkit": "0.2.6",
-    "@solarity/zktype": "0.3.1",
+    "@solarity/zkit": "0.3.0",
+    "@solarity/zktype": "0.4.0-rc.0",
```
Comment (Arvolear) [inline package.json:63]: Let's publish a stable version

File: src/config/config.ts [hunk 2]
```diff
@@ -19,7 +19,7 @@ const defaultConfig: ZKitConfig = {
   setupSettings: {
     contributionSettings: {
       provingSystem: "groth16",
-      contributions: 1,
+      contributions: 2,
```
Comment (Arvolear) [inline src/config/config.ts:22]: + in readme

File: src/core/compile/CompilationProcessor.ts [hunk 3]
```diff
@@ -67,8 +67,9 @@ export class CompilationProcessor {
    * 6. Saves the artifact information using {@link ICircuitArtifacts | CircuitArtifacts}
    *
    * @param filesInfoToCompile Information about circuit files needed for compilation
+   * @param provingSystems An array of {@link ProvingSystemType | proving systems} from the plugin config
    */
-  public async compile(filesInfoToCompile: CircomResolvedFileInfo[]) {
+  public async compile(filesInfoToCompile: CircomResolvedFileInfo[], provingSystems: ProvingSystemType[]) {
```
Comment (Arvolear) [inline src/core/compile/CompilationProcessor.ts:72]: Why do you need to pass `provingSystems` in the compilation job?

File: src/core/zkit/CircuitZKitBuilder.ts [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+  ): Promise<CircuitZKit<ProvingSystemType>> {
+    const circuitArtifact: CircuitArtifact = await this._hre.zkit.circuitArtifacts.readCircuitArtifact(circuitName);
+    const circuitZKitConfig: CircuitZKitConfig = this._getCircuitZKitConfig(circuitArtifact, verifiersDir);
+
+    if (this._isTSProject) {
+      if (circuitArtifact.baseCircuitInfo.protocol.length > 1 && !provingSystem) {
+        throw new HardhatZKitError(
+          "Found several proving systems. Please specify the exact proving system in getCircuit function",
+        );
+      } else if (circuitArtifact.baseCircuitInfo.protocol.length === 1) {
```
Comment (Arvolear) [inline src/core/zkit/CircuitZKitBuilder.ts:49]: Can go without `else`

File: src/core/zkit/CircuitZKitBuilder.ts [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+        const existingProvingSystem: ProvingSystemType = circuitArtifact.baseCircuitInfo.protocol[0];
+
+        if (provingSystem && provingSystem !== existingProvingSystem) {
+          throw new HardhatZKitError(
+            `Passed invalid proving system. Please pass ${existingProvingSystem} proving system or run make task with needed proving systems`,
```
Comment (Arvolear) [inline src/core/zkit/CircuitZKitBuilder.ts:54]:
```suggestion
            `Invalid proving system is passed. Please pass ${existingProvingSystem} proving system or recompile the circuits with the needed one`,
```

File: src/core/zkit/CircuitZKitBuilder.ts [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        provingSystem = undefined;
+      }
+
+      const module = await this._typesGenerator.getCircuitObject(circuitName, provingSystem);
+
+      return new module(circuitZKitConfig);
+    } else {
+      if (!provingSystem || !circuitArtifact.baseCircuitInfo.protocol.includes(provingSystem)) {
+        throw new HardhatZKitError(
+          "Passed undefined or invalid proving system. Please run make task or change proving system.",
```
Comment (Arvolear) [inline src/core/zkit/CircuitZKitBuilder.ts:67]:
```suggestion
          "Undefined or invalid proving system is passed. Please recompile the circuits or change the proving system.",
```

File: src/core/zkit/CircuitZKitBuilder.ts [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+    provingSystem?: ProvingSystemType,
+    verifiersDir?: string,
+  ): Promise<CircuitZKit<ProvingSystemType>> {
+    const circuitArtifact: CircuitArtifact = await this._hre.zkit.circuitArtifacts.readCircuitArtifact(circuitName);
+    const circuitZKitConfig: CircuitZKitConfig = this._getCircuitZKitConfig(circuitArtifact, verifiersDir);
+
+    if (this._isTSProject) {
+      if (circuitArtifact.baseCircuitInfo.protocol.length > 1 && !provingSystem) {
+        throw new HardhatZKitError(
+          "Found several proving systems. Please specify the exact proving system in getCircuit function",
```
Comment (Arvolear) [inline src/core/zkit/CircuitZKitBuilder.ts:47]:
```suggestion
          "Found several proving systems. Please specify the exact proving system in the getCircuit function.",
```

File: src/core/zkit/CircuitZKitBuilder.ts [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+  public async getCircuitZKit(
+    circuitName: string,
+    provingSystem?: ProvingSystemType,
+    verifiersDir?: string,
+  ): Promise<CircuitZKit<ProvingSystemType>> {
+    const circuitArtifact: CircuitArtifact = await this._hre.zkit.circuitArtifacts.readCircuitArtifact(circuitName);
+    const circuitZKitConfig: CircuitZKitConfig = this._getCircuitZKitConfig(circuitArtifact, verifiersDir);
+
+    if (this._isTSProject) {
+      if (circuitArtifact.baseCircuitInfo.protocol.length > 1 && !provingSystem) {
```
Comment (Arvolear) [inline src/core/zkit/CircuitZKitBuilder.ts:45]: Is it actually possible to get into this `if`?

File: src/index.ts [hunk 9]
```diff
@@ -182,17 +187,34 @@ const setup: ActionType<SetupTaskConfig> = async (taskArgs: SetupTaskConfig, env
...
     for (const setupInfo of circuitSetupInfoArr) {
+      const currentSetupCacheEntry = CircuitsSetupCache!.getEntry(setupInfo.circuitArtifactFullPath);
+
+      let currentProvingSystemsData: ProvingSystemData[] = [];
+
+      if (currentSetupCacheEntry) {
+        currentProvingSystemsData = currentSetupCacheEntry.provingSystemsData.filter((data: ProvingSystemData) => {
+          return !setupInfo.provingSystems.includes(data.provingSystem);
+        });
+      }
```
Comment (Arvolear) [inline src/index.ts:203]: Why do you need to filter here?

File: src/index.ts [hunk 10]
```diff
@@ -288,44 +317,6 @@ const clean: ActionType<any> = async (_taskArgs: any, env: HardhatRuntimeEnviron
   fs.rmSync(circuitTypesFullPath, { recursive: true, force: true });
 };
 
-const getCircuitZKit: ActionType<GetCircuitZKitConfig> = async (
```
Comment (Arvolear) [inline src/index.ts:291]: Why not still keep it for clarity?

Comment (Hrom131) [reply src/index.ts:291]: Agree

File: src/index.ts [hunk 11]
```diff
@@ -370,3 +388,9 @@ zkitScope
   .setAction(generateVerifiers);
 
 zkitScope.task(TASK_ZKIT_CLEAN, "Clean all circuit artifacts, keys, types and etc").setAction(clean);
+
+subtask(SUBTASK_ZKIT_GET_CIRCUIT_ZKIT)
+  .addOptionalParam("verifiersDir", undefined, undefined, types.string)
+  .addOptionalParam("verifierTemplateType", undefined, undefined, types.any)
+  .addParam("circuitName", undefined, undefined, types.string)
```
Comment (Arvolear) [inline src/index.ts:395]: This list is outdated

File: src/utils/constraints-utils.ts [hunk 12]
```diff
@@ -0,0 +1,93 @@
+import * as snarkjs from "snarkjs";
+// @ts-expect-error: No type definitions available for the "r1csfile" package
+import * as r1csfile from "r1csfile";
+// @ts-expect-error: No type definitions available for the "@iden3/binfileutils" package
+import * as binfileutils from "@iden3/binfileutils";
```
Comment (Arvolear) [inline src/utils/constraints-utils.ts:5]: Do we have this package in the deps?

File: src/utils/constraints-utils.ts [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+
+import { LinearCombination, R1CSConstraint } from "../../src/types/utils";
+
+export async function getGroth16ConstraintsNumber(r1csFilePath: string): Promise<number> {
```
Comment (Arvolear) [inline src/utils/constraints-utils.ts:9]: Let's name it just `getR1CSConstraintsNumber`

File: src/utils/constraints-utils.ts [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+    });
+  };
+
+  const join = (lc1: LinearCombination, k: bigint, lc2: LinearCombination) => {
+    const res = { ...lc1 };
+    Object.keys(lc2).forEach((s) => {
+      res[s] = res[s] ? Fr.add(res[s], Fr.mul(k, lc2[s])) : lc2[s];
+    });
+    normalize(res);
+    return res;
```
Comment (Arvolear) [inline src/utils/constraints-utils.ts:26]:
```suggestion
    const res = { ...lc1 };
    
    Object.keys(lc2).forEach((s) => {
      res[s] = res[s] ? Fr.add(res[s], Fr.mul(k, lc2[s])) : lc2[s];
    });
    
    normalize(res);
    
    return res;
```

File: src/utils/constraints-utils.ts [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+    let n = 0;
+
+    Object.keys(lc).forEach((key) => {
+      if (lc[key] !== 0n) {
+        if (key === "0") {
+          k = Fr.add(k, lc[key]);
+        } else {
+          n++;
+        }
+      }
```
Comment (Arvolear) [inline src/utils/constraints-utils.ts:54]: Shitty nested `if`...

```suggestion
      if (lc[key] === 0n) {
        return;
      }

      if (key === "0") {
        k = Fr.add(k, lc[key]);
      } else {
          n++;
      }
```

File: src/core/compile/CompilationProcessor.ts [hunk 16]
```diff
@@ -11,6 +10,8 @@ import { CircomCompilerFactory, createCircomCompilerFactory, getHighestVersion,
 import { HardhatZKitError } from "../../errors";
 import { CIRCUIT_ARTIFACT_VERSION, NODE_MODULES } from "../../constants";
 import { Reporter } from "../../reporter";
+
+import { getGroth16ConstraintsNumber } from "../../utils/constraints-utils";
 import { getNormalizedFullPath, renameFilesRecursively, readDirRecursively } from "../../utils/path-utils";
```
Comment (Hrom131) [inline src/core/compile/CompilationProcessor.ts:15]: We can merge these imports

File: src/utils/constraints-utils.ts (hunk 12 above)
Comment (mllwchrry) [reply src/utils/constraints-utils.ts:5]: no, but they are inside snarkjs dependencies

Comment (Arvolear) [reply src/utils/constraints-utils.ts:5]: By why do we need to rely on some transient dependencies?

**Pull request merged**